### PR TITLE
Fix PP clip_grad_norm

### PR DESCRIPTION
**Pull request by zijian-hu**
This PR fixes the pipeline parallel incomplete gradient norm issue. Refer to https://github.com/pytorch/torchtitan/issues/596 for more details.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -307,8 +308,11 @@ def loss_fn(pred, labels):
 
             # clip gradients
             for m in model_parts:
```
Comment (tianyu-l) [inline train.py]: I think we shouldn't do this for each m in model_parts. The `total_norm` has to be an overall sum.

**Review by tianyu-l — COMMENTED** [review]
It may not be easy / feasible, but can we find a way to
1. first get global `total_norm`
2. then make a function call to (some variant of) the `torch.nn.utils.clip_grad_norm_` with the global `total_norm`

The idea is to avoid the largely repeated code in pytorch.

File: train.py (hunk 1 above)
Comment (zijian-hu) [reply train.py]: You are right. This for loop would introduce more problems. If my understanding is correct, `model_parts` contains more than 1 nn.Module only when we are using multi-stage PP schedules. So we can simply just chain the `.parameters()` then pass to the clip norm function

Comment (zijian-hu) [reply train.py]: Pushed a new commit with the above mentioned logic

File: torchtitan/clip_grad_nrom.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    if isinstance(total_norm, DTensor):
+        # if total_norm is a DTensor, the placements must be `torch.distributed._tensor.ops.math_ops._NormPartial`
+        # we can simply reduce the DTensor to get the total norm in this tensor's process group
+        # and then convert it to a local tensor
+        total_norm = total_norm.redistribute(
+            placements=[Replicate()] * total_norm.device_mesh.ndim
+        ).to_local()
+
+    total_norm **= norm_type
+    dist.all_reduce(total_norm, op=dist.ReduceOp.SUM, group=pp_mesh.get_group())
```
Comment (wconstab) [inline torchtitan/clip_grad_nrom.py]: I don't like that we have to copy/modify the whole clip_grad_norm, and I think I see the reason why we do it and might have an idea for avoiding it.

IIUC the issue is that clip_grad_norm bundles 2 actions into one API.
1. compute total norm
2. apply clipping based on total norm

would it be cleaner if we provided unbundled APIs at the torch layer for these 2?  then in the PP case we can just call the first helper, do the PP reduce, then call the second helper.

Comment (tianyu-l) [reply torchtitan/clip_grad_nrom.py]: exactly what we need!

Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py]: It would be great if `torch.nn.utils.clip_grad_norm_` can be split into 2 APIs

Comment (awgu) [reply torchtitan/clip_grad_nrom.py]: I think we can ask core team. It feels reasonable to split the total norm computation from the actual clipping.

Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py]: @awgu should we create a github issue/feature request under pytorch main repo?

Comment (awgu) [reply torchtitan/clip_grad_nrom.py]: Yes let us do that. I can ping people internally after we make the issue too.

Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py]: @awgu issue in pytorch main repo created. See https://github.com/pytorch/pytorch/issues/139467

Comment (H-Huang) [reply torchtitan/clip_grad_nrom.py]: Hi @zijian-hu, thank you for your changes so far. It looks like https://github.com/pytorch/pytorch/pull/139662 is getting checked in, could you update your PR to use these APIs? Thanks so much!

Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py]: @H-Huang Just saw the pytorch issue has been resolved. Will update this PR ASAP

File: train.py (hunk 1 above)
Comment (zijian-hu) [reply train.py]: Updated with the newest PyTorch clip norm APIs

File: torchtitan/clip_grad_nrom.py (hunk 2 above)
Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py]: @H-Huang Updated with the newest PyTorch clip norm APIs

File: torchtitan/clip_grad_nrom.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        pp_mesh: pipeline parallel device mesh. If not None, will reduce gradient norm across PP stages.
+
+    Returns:
+        Total norm of the parameter gradients (viewed as a single vector).
+
+    """
+
+    total_norm = torch.nn.utils.get_total_norm(
+        parameters, norm_type, error_if_nonfinite, foreach
+    )
```
Comment (awgu) [inline torchtitan/clip_grad_nrom.py]: We want total norm of gradients not parameters:
```suggestion
    grads = [p.grad if p.grad is not None for p in parameters]
    total_norm = torch.nn.utils.get_total_norm(
        grads, norm_type, error_if_nonfinite, foreach
    )
```

File: torchtitan/clip_grad_nrom.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        # if total_norm is a DTensor, the placements must be `torch.distributed._tensor.ops.math_ops._NormPartial`
+        # we can simply reduce the DTensor to get the total norm in this tensor's process group
+        # and then convert it to a local tensor
+        total_norm = total_norm.redistribute(
+            placements=[Replicate()] * total_norm.device_mesh.ndim
+        ).to_local()
+
+        total_norm **= norm_type
+        dist.all_reduce(total_norm, op=dist.ReduceOp.SUM, group=pp_mesh.get_group())
+        total_norm **= 1.0 / norm_type
```
Comment (awgu) [inline torchtitan/clip_grad_nrom.py]: This is only correct for `p`-norm where `p` is `int`/`float` (e.g. this does not work for `inf`-norm).

Intuitively, we should somehow allow us to change `total_norm` to have `_NormPartial(norm_type)` placement over the PP mesh, and we should redistribute it to `Replicate()` to reuse DTensor's own logic. cc: @tianyu-l @wz337 @XilunWu  is there any way to do this?

**Review by awgu — CHANGES_REQUESTED** [review]
sorry requesting changes since `total_norm` is computed wrt. parameters instead of gradients

File: torchtitan/clip_grad_nrom.py (hunk 4 above)
Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py]: @awgu in torchtitan, PP is done by first manually splitting the Llama model then wrap in PP wrapper. If my understanding is correct, none of them converts models weights into DTensor with PP mesh placement

File: torchtitan/clip_grad_nrom.py
Comment (tianyu-l) [inline torchtitan/clip_grad_nrom.py:1]: given the function is now simpler, can be put it in `torchtitan/utils.py`?

Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py:1]: Sure. Will be moved to `torchtitan/utils.py`

File: torchtitan/clip_grad_nrom.py (hunk 4 above)
Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py]: For now, will just add a if-statement to check for inf norm

Comment (wconstab) [reply torchtitan/clip_grad_nrom.py]: > PP is done by first manually splitting the Llama model then wrap in PP wrapper. If my understanding is correct, none of them converts models weights into DTensor with PP mesh placement

this is true, however after splitting the llama model we do also apply SPMD parallelisms (TP and DP) which will use DTensor for parameters.  So by the time we are clipping norms, parameters would be DTensors.  (if we use PP only and no TP/DP, then there would be no DTensor parameters)

Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py]: @wconstab You are right about SPMD (TP and/or DP) but the DTensor in that case has placement on TP/DP mesh not on PP mesh. We still need to do some form of reduce either explicitly (same as my code) or wrap the `total_norm` in DTensor on PP mesh (maybe via `DTensor.redistribute`?)

File: torchtitan/clip_grad_nrom.py
Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py:1]: Moved to `torchtitan/utils.py`

File: torchtitan/clip_grad_nrom.py (hunk 3 above)
Comment (zijian-hu) [reply torchtitan/clip_grad_nrom.py]: Fixed. PS the suggested code has a tiny bug.

File: torchtitan/utils.py [hunk 5]
```diff
@@ -263,3 +266,63 @@ class NoColor:
...
+    Returns:
+        Total norm of the parameter gradients (viewed as a single vector).
+
+    """
+    grads = [p.grad for p in parameters if p.grad is not None]
+    total_norm = torch.nn.utils.get_total_norm(
+        grads, norm_type, error_if_nonfinite, foreach
+    )
+
+    if pp_mesh is not None and isinstance(total_norm, DTensor):
```
Comment (wconstab) [inline torchtitan/utils.py]: Hm, if we use PP without tp/dado don't we still need to do the allreduce even though there are no Dtensor? Should we break this if into 2?

Comment (zijian-hu) [reply torchtitan/utils.py]: You are right. We need to support PP only case. The if statement should be set to nested if's. The code has been updated

File: torchtitan/utils.py [hunk 6]
```diff
@@ -263,3 +266,66 @@ class NoColor:
...
+
+    """
+    grads = [p.grad for p in parameters if p.grad is not None]
+    total_norm = torch.nn.utils.get_total_norm(
+        grads, norm_type, error_if_nonfinite, foreach
+    )
+
+    if pp_mesh is not None:
+        if isinstance(total_norm, DTensor):
+            # will reach here if PP + other parallelism is used. If only using PP, total_norm will be a local tensor
```
Comment (zijian-hu) [inline torchtitan/utils.py:313]: @wconstab the if statement are now separated. Please let me know if it make sense to you

Comment (wconstab) [reply torchtitan/utils.py:313]: It looks good, thanks!

File: torchtitan/utils.py [hunk 7]
```diff
@@ -263,3 +266,66 @@ class NoColor:
...
+
+            # if total_norm is a DTensor, the placements must be `torch.distributed._tensor.ops.math_ops._NormPartial`
+            # we can simply reduce the DTensor to get the total norm in this tensor's process group
+            # and then convert it to a local tensor
+            total_norm = total_norm.redistribute(
```
Comment (H-Huang) [inline torchtitan/utils.py]: nit:  use `total_norm.full_tensor()` instead?

**Review by H-Huang — APPROVED** [review]
Looks good to me!

File: torchtitan/utils.py (hunk 7 above)
Comment (zijian-hu) [reply torchtitan/utils.py]: @H-Huang Good suggestion. Updated!

**Comment by zijian-hu** [comment]
> sorry requesting changes since `total_norm` is computed wrt. parameters instead of gradients

@awgu please let me know if the update make sense and feel free to resolve the change request if you believe it is good enough

**Review by awgu — APPROVED** [review]
sounds good!
